### PR TITLE
Entry verify service

### PR DESCRIPTION
**Pull request by sakridge**
#### Problem

Poh verify of entries taking too long.

#### Summary of Changes

Prioritize slots to verify.

Fixes #

### CODE REVIEW

File: ledger/src/entry_verify_service.rs [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+                .unwrap()
+                .parent()
+                .unwrap();
+            let parent_slot = parent.slot();
+            let parent_hash = parent.hash();
+            unverified_blocks.add_unverified_block(slot, parent_slot, entries, weight, parent_hash);
+        }
+        if let Some(heaviest_leaf) = unverified_blocks.next_heaviest_leaf() {
+            let heaviest_ancestors = unverified_blocks.get_unverified_ancestors(heaviest_leaf);
+            if let Some(heaviest_slot) = heaviest_ancestors.iter().next() {
```
Comment (carllin) [inline ledger/src/entry_verify_service.rs:76]: I think we should read from `slot_receiver` after verifying every bank, and make sure the fork we're working on is still the heaviest. That way we don't get stuck verifying an entire branch before updating the weights.

Comment (sakridge) [reply ledger/src/entry_verify_service.rs:76]: Yea.. it's what it's doing, no?

Comment (carllin) [reply ledger/src/entry_verify_service.rs:76]: ah yeah, i saw the word `iter` and assumed it was a loop xD

File: ledger/src/entry_verify_service.rs [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        bank_forks: &Arc<RwLock<BankForks>>,
+        slot_verify_results: &Arc<RwLock<HashMap<Slot, bool>>>,
+        unverified_blocks: &mut UnverifiedBlocks,
+    ) -> Result<(), RecvTimeoutError> {
+        let root_slot = bank_forks.read().unwrap().root();
+
+        unverified_blocks.set_root(root_slot);
+
+        while let Ok((slot, entries, weight)) = slot_receiver.try_recv() {
+            let parent = bank_forks
```
Comment (carllin) [inline ledger/src/entry_verify_service.rs:63]: if the sender also sends the parent, I think we can avoid this lock here

Comment (carllin) [inline ledger/src/entry_verify_service.rs:63]: small nit: I think if sender sends parent bank retrieved through `bank.parent()` we can avoid lock here.

File: ledger/src/entry_verify_service.rs [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+        }
+        if let Some(heaviest_leaf) = unverified_blocks.next_heaviest_leaf() {
+            let heaviest_ancestors = unverified_blocks.get_unverified_ancestors(heaviest_leaf);
+            if let Some(heaviest_slot) = heaviest_ancestors.iter().next() {
+                let block_info = unverified_blocks
+                    .unverified_blocks
+                    .get(heaviest_slot)
+                    .unwrap();
+                let verify_result = block_info.entries.verify(&block_info.parent_hash);
+                slot_verify_results
```
Comment (carllin) [inline ledger/src/entry_verify_service.rs:82]: Does this object also get cleaned up when we set a new root?

File: ledger/src/entry_verify_service.rs [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+            let parent = bank_forks
+                .read()
+                .unwrap()
+                .get(slot)
+                .unwrap()
+                .parent()
+                .unwrap();
+            let parent_slot = parent.slot();
+            let parent_hash = parent.hash();
+            unverified_blocks.add_unverified_block(slot, parent_slot, entries, weight, parent_hash);
```
Comment (carllin) [inline ledger/src/entry_verify_service.rs:72]: In this `try_recv` can we read out all the pending items in the channel  and insert them all through `add_unverified_block` before calling `next_heaviest_leaf`? That way we get the overall best before committing to work

File: core/src/progress_map.rs [hunk 5]
```diff
@@ -293,6 +297,38 @@ impl std::ops::DerefMut for ProgressMap {
 }
 
 impl ProgressMap {
+    pub fn add_new_progress_entry(
```
Comment (carllin) [inline core/src/progress_map.rs:300]: Factored out from replay_stage `replay_active_banks`

**Pull request closed without merging**